clang/test/SemaSYCL/attr-syclglobalvar.cpp

```diff
@@ -0,0 +1,54 @@
+// RUN: %clang_cc1 -fsycl-is-device -verify -fsyntax-only %s
+
+#include "Inputs/sycl.hpp"
+
+__attribute__((sycl_global_var)) int GlobalWithAttribute;
+
+__attribute__((sycl_global_var)) extern int ExternGlobalWithAttribute;
+
+namespace NS {
+    __attribute__((sycl_global_var)) int NSGlobalWithAttribute;
+}
+
+union U {
+    int InstanceMember;
+};
+
+__attribute__((sycl_global_var)) U GlobalUnion;
+
+struct S {
+    __attribute__((sycl_global_var)) static int StaticMember;
+
+    // expected-warning@+1 {{attribute only applies to global variables}}
+    __attribute__((sycl_global_var)) int InstanceMember;
+};
+int S::StaticMember = 0;
+
+__attribute__((sycl_global_var)) S GlobalStruct;
+
+__attribute__((sycl_global_var)) static S StaticGlobal;
+
+// expected-error@+1 {{attribute takes no arguments}}
+__attribute__((sycl_global_var(42))) int GlobalWithAttributeArg;
+
+int GlobalNoAttribute;
+
+// expected-warning@+1 {{attribute only applies to global variables}}
+__attribute__((sycl_global_var)) void F() {
+    __attribute__((sycl_global_var)) static int StaticLocalVar;
```

Comment:
I would have expected a diagnostic here as a local variable is not a global variable.

Reply:
> I would have expected a diagnostic here as a local variable is not a global variable.

This is an interesting question.  SYCL spec applies the "global variable" restriction (that it must be const) to variables of static storage duration.
Here: https://www.khronos.org/registry/SYCL/specs/sycl-2020/html/sycl-2020.html#sec:language.restrictions.kernels
and here: https://www.khronos.org/registry/SYCL/specs/sycl-2020/html/sycl-2020.html#subsec:commonAddressSpace

Given this, I wonder if the attribute could be applied to the file-scope static variables to override the restriction.

Reply:
I updated the tests and added the anonymous static union.

The anonymous union is not considered a global (doesn't accept the attribute) *and* isn't allowed in a kernel.

The static local var, surprisingly, is considered a global and accepts the attribute, though still not allowed in a kernel (because I didn't use the attribute's presence to allow that specific case).

These are interesting results. In my changes, I relied on the pre-defined "GlobalVar" subject list. "GlobalVar" is defined earlier as a Var where the code check `S->hasGlobalStorage()` is true.

I'm stepping through code so I can answer why some of these results "have global storage" and some don't.

Reply:
Ultimately, though, this isn't a feature we're expecting external users to use. This is meant to be used by *us* to enable the assert feature. Perhaps there's a different subject list I can use that applies to fewer things?

Reply:
```
  /// Returns true for all variables that do not have local storage.
  ///
  /// This includes all global variables as well as static variables declared
  /// within a function.
  bool hasGlobalStorage() const { return !hasLocalStorage(); }
```
That's why you get the behavior you're getting (and the name `GlobalVar` in Attr.td is *awful*, IMHO).

> Ultimately, though, this isn't a feature we're expecting external users to use. This is meant to be used by us to enable the assert feature. Perhaps there's a different subject list I can use that applies to fewer things?

It doesn't much matter what we expect -- once we expose the attribute, users will use the attribute. This is basically giving users a blessed way to opt-in to using global variables whenever they'd like (I'm not convinced this is a good idea, btw) because we found a use case where we want to do it ourselves. We can give this attribute whatever semantics we'd like, so there may be a more restrictive predicate we can use, but without a design document for the feature as such, I don't really know what the right answer is as a reviewer (I can't go off the SYCL 2020 spec because this attribute is intended specifically to do things the SYCL spec doesn't allow).

Reply:
I think the answer boils down to trying to figure out what we were protecting access to and why. If it's the *storage* we care about, then `hasGlobalStorage()` is the right answer. If it's the *name of the variable* we care about, then maybe you'd want something more like `!Decl::isLocalVarDecl()`.

I may be wrong, but the impression I have from the SYCL spec is that the storage is the issue, but it's not fully clear because some storage seems to be fine (such as `const`).

Reply:
I suppose we're protecting users from writing code that *appears* to work on host but would break on GPU or other device. So this attribute becomes one of those "trust me, I'll be careful" opt-in kind of features. For general purposes, storage is what matters, even if the name for that storage has a smaller scope. For *our* purposes, we care about both global storage and global name, since we intend to use this global variable from multiple TUs. I'll add !isLocalVarDecl to the check to see if I can reduce the surface area of this feature.

Reply:
> So this attribute becomes one of those "trust me, I'll be careful" opt-in kind of features. 

TBH, I think this one is a bit worse than that -- it's "I want to ignore the standard requirements here".

> For our purposes, we care about both global storage and global name, since we intend to use this global variable from multiple TUs. I'll add !isLocalVarDecl to the check to see if I can reduce the surface area of this feature.

I was thinking about this a bit over the weekend and here's a brain dump of what I was thinking.

I don't think this functionality should exist as a user-facing attribute. It's basically hacking around the standard requirements because of a special case. As a general rule, if we *have to* add a hack to ignore specification requirements, it should be something that's quite clearly identified as not-fit-for-general use by making everyone feel somewhat uncomfortable with its existence, and it should be difficult for users to use it without resorting to hacks of their own.

I sort of wonder if the way to do this is to name the attribute `sycl_global_var_HACK_FOR_SYCL_HEADERS_ONLY` and then diagnose any use of the attribute outside of a system header. This way, users can't use it easily without hacks of their own (they have to pretend their code is a system header to get it to compile) and if they do go to the trouble, they see a big, ugly name that should make them feel bad. (Then we'd remove the docs from AttrDocs.td as well, obviously.)

Reply:
Sounds reasonable to me. @s-kanaev Thoughts? I think the only difference from your perspective would be the longer attribute name, "sycl_global_var_HACK_FOR_SYCL_HEADERS_ONLY".

Reply:
> This way, users can't use it easily without hacks of their own (they have to pretend their code is a system header to get it to compile)

Will it break compilation of the user's code which includes DPCPP RT header with declaration and/or definition of such a variable?

Reply:
@s-kanaev -- if the code includes the header as a system header, then everything will work fine. Examples of what won't work fine are:

* Writing the attribute in user source code (obviously, that's the point)
* Including SYCL as something other than a system header. I don't know why someone would want to do that and it seems like something we should not support anyway.
* Including the SYCL header in a source file, preprocessing that to a file without line markers, then compiling that preprocessed code (if the preprocessed code keeps the line markers, everything is fine, but if the line markers are removed, then it won't compile). Line markers are kept by default, so the user is unlikely to do this by accident.
* Copying code out of the SYCL headers and into your own files. Works in SYCL, same code fails in your own file.

Reply:
Sounds good.

Reply:
Latest commit expands the test to distinguish between user code and system-include code. Currently, the attribute is still allowed to *appear* in user code, but it won't have any effect. The attribute will only have an effect if it's used from a system header. To disallow the attribute from even *appearing* in user code, I'll need to figure out how to do a source location check from within a tablegen subject list.

Reply:
On the second thought, here is another example:
```c++
// sycl.hpp, system header
template <typename T>
struct Struct {
  __attribute__((sycl_global_var)) static T Variable;
};

// main.cpp, user's application
struct UserTy {
  int X;
  int Y;
};

Struct<UserTy> Global;
```

Is this use-case allowed?

Reply:
I would imagine that use case should be allowed -- the declaration of the variable is appropriately marked, the fact that the type comes from a user instantiation shouldn't matter to the attribute's semantics.

Reply:
Latest commit updated with non-simple handler. Also included templates in tests. `Struct<UserTy>::StaticMember` would work, but `Struct<UserTy> Global;` would not because you can't use a global in the kernel, and you can't mark a global in the user's cpp file (only system headers).